src/lib/Inference.hs

```diff
@@ -1567,15 +1571,35 @@ binderNestAsPiNest arr = \case
   Empty             -> Empty
   Nest (b:>ty) rest -> Nest (PiBinder b ty arr) $ binderNestAsPiNest arr rest
 
+inferRoleBinderNest :: Fallible m => Nest Binder n l -> m (Nest RoleBinder n l)
+inferRoleBinderNest bs = forEachNestItemM bs \(b:>ty) -> do
+  role <- inferRole ty PlainArrow
+  return $ RoleBinder b ty role
+{-# INLINE inferRoleBinderNest #-}
+
+inferRolePiBinderNest :: Fallible m => Nest PiBinder n l -> m (Nest RolePiBinder n l)
+inferRolePiBinderNest bs = forEachNestItemM bs \(PiBinder b ty arr) -> do
+  role <- inferRole ty arr
+  return $ RolePiBinder b ty arr role
+{-# INLINE inferRolePiBinderNest #-}
+
+inferRole :: Fallible m => Type n -> Arrow -> m ParamRole
+inferRole ty arr = case (ty, arr) of
+  (TyKind  , _         ) -> return TypeParam
+  (DictTy _, ClassArrow) -> return DictParam
```

Comment:
Actually isn't anything that uses a ClassArrow a DictParam? The treatment of DictParams is mainly based on the way they are generated (i.e. through synthesis), not about them having a dict type? I don't think the code as written is incorrect, but perhaps a little confusing due to a too strong precondition.

Reply:
Yeah, the more I think about it, the more I am convinced that actually the role inference might be better rephrased as:
```haskell
case arr of
  ClassArrow -> DictParam
  _ -> isData ty >>= \case
    True -> DataParam
    False -> TypeParam
```
WDYT?